tutorials/misc/state_design_pattern.rst

```diff
@@ -0,0 +1,254 @@
+script will be an implementation of a different state.
+
+.. note::
+  There is a great resource explaining the concept of the state design pattern here:
+  https://gameprogrammingpatterns.com/state.html
+
+Script-Setup
+------------
+
+The feature of inheritance is useful for getting started with this design principle.
+A class should be created that describes the base features of the player. For now, a
+player will be limited to two actions: **move left**, **move right**. This means
+there will be two states: **idle** and **run**.
+
+Below is the generic state, from which all other states will inherit.
+
+.. tabs::
+  .. code-tab:: gdscript GDScript
+
+    extends Node2D
+    
+    class_name State
+
+    var change_state
+    var animated_sprite
+    var persistent_state
+    var velocity
+
+    func _physics_process(delta):
+      persistent_state.move_and_slide(persistent_state.velocity, Vector2.UP)
+
+    func setup(change_state, animated_sprite, kinematic_body_2d):
+	self.change_state = change_state
+	self.animated_sprite = animated_sprite
+	self.persistent_state = persistent_state
+
+    func move_left():
+	pass
+
+    func move_right():
+	pass
+
+A few notes on the above script. First, this implementation uses a 
+``setup(change_state, animated_sprite, persistent_state)`` method to assign
+references. These references will be instantiated in the parent of this state. This helps with something 
+in programming known as *cohesion*. The state of the player does not want the responsibility of creating 
+these variables, but does want to be able to use them. However, this does make the state *coupled* to the 
+state's parent. This means that the state is highly reliant on whether it has a parent which contains 
+these variables. So, remember that *coupling* and *cohesion* are important concepts when it comes to code management.
+
+.. note:: 
+  See the following page for more details on cohesion and coupling:
+  https://courses.cs.washington.edu/courses/cse403/96sp/coupling-cohesion.html
+
+Second, there are some methods in the script for moving, but no implementation. The state script
+just uses ``pass`` to show that it will not execute any instructions when the methods are called. This is important.
+
+Third, the ``_physics_process(delta)`` method is actually implemented here. This allows the states to have a default
+``_phyics_process(delta)`` implementation where ``velocity`` is used to move the player. The way that the states can modify
+the movement of the player is to use the ``velocity`` variable defined in their base class.
+
+Finally, this script is actually being designated as a class named ``State``. This makes refactoring the code
+easier, since the file path from using the ``load()`` and ``preload()`` functions in godot will not be needed.
+
+So, now that there is a base state, the two states discussed earlier can be implemented.
+
+.. tabs::
+  .. code-tab:: gdscript GDScript
+
+    # filename: idle_state.gd
```

Comment:
It should be enough to just use `# idle_state.gd` without specifying `filename: `. That's what is done in most other tutorials.